core/shared/src/main/scala/zio/internal/metrics/MetricHook.scala

```diff
@@ -18,7 +18,8 @@ package zio.internal.metrics
 
 private[zio] final case class MetricHook[-In, +Out](
   update: In => Unit,
-  get: () => Out
+  get: () => Out,
+  modify: In => Unit
```

Comment:
In general, for some type `In`, what is the meaning of `update`, and what is the meaning of `modify`?

Reply:
That is a good question.

I think the basic idea here is that we can have metrics that can accept more than one type of message. In particular, in the case of a `Gauge` it can accept an `In` that says "set me to this value" or an `In` that says "increment me by this value".

So I think the meaning is that when a metric accepts more than one type of `In` message, `update` ignores the prior metric state (setting the `Gauge`) and `modify` incorporates the prior metric state along with the message into the new state (increment the `Gauge`). For metrics that only accept one type of input message they are identical.

Reply:
I think maybe what we're getting at is that some metrics can be "set" and others can be "updated" and still others can be "set" and "updated".

This almost argues for 2 type parameters: `Absolute`, `Patch`, with two methods: `update` (accepting `Patch`) and `set` (accepting `Absolute`). For some metrics, `Patch` = `Unit` because they don't support patching (e.g. `Counter`). For some metrics, `Absolute` = `Unit` because they don't support setting (`Frequency`).

Does that make sense or not?

Reply:
I think that does conceptually make sense, though it seems like there might be some subjectivity in what is updating and setting. Like I would think a `Counter` supports updating but not setting because the increment creates a new state the reflects the original state and the new input versus just the new input. I think of the existing metrics a `Gauge` is actually the only one that supports a true "set" semantic in that sense.

The other issue is binary compatibility and how we get from here to there which is part of what led me to the current implementation since I think it conceptually does that without changing the current API.